nixos/tests/k3s/containerd-config.nix

```diff
@@ -0,0 +1,58 @@
+# A test that containerdConfigTemplate settings get written to containerd/config.toml
+import ../make-test-python.nix (
+  { lib, k3s, ... }:
+  let
+    nodeName = "test";
+    shutdownGracePeriod = "1m13s";
+    shutdownGracePeriodCriticalPods = "13s";
+  in
+  {
+    name = "${k3s.name}-containerd-config";
+    nodes.machine =
+      { ... }:
+      {
+        # k3s uses enough resources the default vm fails.
+        virtualisation.memorySize = 1536;
+        virtualisation.diskSize = 4096;
+
+        services.k3s = {
+          enable = true;
+          package = k3s;
+          # Slightly reduce resource usage
+          extraFlags = [
+            "--disable coredns"
+            "--disable local-storage"
+            "--disable metrics-server"
+            "--disable servicelb"
+            "--disable traefik"
+            "--node-name ${nodeName}"
+          ];
+          gracefulNodeShutdown = {
+            enable = true;
+            inherit shutdownGracePeriod shutdownGracePeriodCriticalPods;
+          };
+          containerdConfigTemplate = ''
+            # Base K3s config
+            {{ template "base" . }}
+
+            # MAGIC COMMENT
+          '';
+        };
+      };
+
+    testScript = ''
+      start_all()
+      machine.wait_for_unit("k3s")
+      # wait until the node is ready
+      machine.wait_until_succeeds(r"""kubectl wait --for='jsonpath={.status.conditions[?(@.type=="Ready")].status}=True' nodes/${nodeName}""")
+      # test whether the config template file contains the magic comment
+      out=machine.succeed("cat /var/lib/rancher/k3s/agent/etc/containerd/config.toml.tmpl")
+      assert "MAGIC COMMENT" in out, "the containerd config template does not contain the magic comment"
+      # test whether the config file contains the magic comment
+      out=machine.succeed("cat /var/lib/rancher/k3s/agent/etc/containerd/config.toml")
+      assert "MAGIC COMMENT" in out, "the containerd config does not contain the magic comment"
```

Comment:
This is testing the implementation, not the effect. I think a better practice is to test that if we actually configure something meaningful, we make sure k3s behaves as expected.

Reply:
K3s picks up the `config.toml.tmpl` file and fills in the template to produce `config.toml`. To me that is an "effect" which shows that K3s is receiving the config.

We could test something further but I'm not sure what. I'm using this at the moment to install new container runtimes, which would require a fair bit more complexity to test here.

Reply:
Can you test configuring some `containerd` feature to show `containerd` is actually this configuration file?

Reply:
> K3s picks up the `config.toml.tmpl` file and fills in the template to produce `config.toml`. To me that is an "effect" which shows that K3s is receiving the config.

I didn't think about that. Thanks!

Reply:
@thomasjm We could test adding container runtimes too. If you are willing, obviously. That way it helps guarantee upgrades won't be silently breaking your use case.

Reply:
If it's okay with you all, I think I'd like to leave this as-is for now -- this is only step 1 of my master plan to enable Kata Containers. That work will test all this in an end-to-end way.